### DISCUSSION TITLE
TorchMetrics wrapper around another library's metrics

### DISCUSSION THREAD

**Discussion by ibro45** [Lightning-AI/torchmetrics#2263] [Q&A] [answered]
Hi!

I would like to have a TorchMetrics wrapper around [MONAI metrics](https://docs.monai.io/en/latest/metrics.html). I initially implemented this, but I realized that it does not work properly, and I assume it is because of lack of `.add_state()`:

```python
import torchmetrics


class MonaiToTorchMetricsWrapper(torchmetrics.Metric):
    def __init__(self, monai_metric_class, **kwargs):
        super().__init__()
        self.monai_metric = monai_metric_class(**kwargs)

    def update(self, preds, target):
        return self.monai_metric(preds, target)

    def compute(self):
        return self.monai_metric.aggregate()

    def reset(self):
        self.monai_metric.reset()

```

How would you go about doing that?

**Comment by ibro45** [comment]
Here's exactly what's happening:

- First row - torchmetrics Dice
- Second row - MONAI Dice
- On the right - metric at a particular step. The two metrics are almost the same, there are some differences due to different implementations.
- On the left - metric over an epoch. You can see that the torchmetrics and MONAI Dice differ here. Specifically, MONAI Dice is not corrected correctly for the epoch. Note, that the MONAI Dice is wrapped into a torchmetrics Metric, as posted above.

**Validation**
<img width="942" alt="image" src="https://github.com/Lightning-AI/torchmetrics/assets/18015788/8f521449-1d7c-46b8-ac54-c8a1f878ee28">

**Training**
<img width="942" alt="image" src="https://github.com/Lightning-AI/torchmetrics/assets/18015788/663ab892-e7ee-4a7c-9907-c7f8aa81caa0">

**Comment by nkaenzig** [comment]
One issue in your wrapper class is that in the `.reset()` method, you should also call torchmetric's reset:

```python
    def reset(self):
        super().reset()
        self. monai_metric.reset()
```

Also make sure to configure aggregations / reductions correctly. I think monai's implementation using `reduction="mean", ignore_empty=False` is equivalent to using `average="macro"` in  `torchmetrics.Dice` (`"micro"` is the default there).

**Comment by Borda** [comment, marked as answer]
@nkaenzig is right about calling `super().reset()` — that's essential. Let me provide a complete working pattern for wrapping external metrics (like MONAI's) so they work correctly with Lightning's epoch-level aggregation.

The core issue is: without `add_state`, TorchMetrics can't sync or aggregate anything across batches/devices. You need to accumulate raw values yourself.

**Complete pattern:**
```python
import torch
from torchmetrics import Metric
from torchmetrics.utilities import dim_zero_cat


class WrappedExternalMetric(Metric):
    """Wrap any external metric that takes (preds, target) -> scalar."""

    is_differentiable = False
    higher_is_better = True  # adjust per metric

    def __init__(self, external_fn, **kwargs):
        super().__init__(**kwargs)
        self.external_fn = external_fn
        # Store per-batch scores; "cat" merges across distributed processes
        self.add_state("scores", default=[], dist_reduce_fx="cat")

    def update(self, preds: torch.Tensor, target: torch.Tensor) -> None:
        # Compute the external metric on this batch
        score = self.external_fn(preds, target)
        if not isinstance(score, torch.Tensor):
            score = torch.tensor(score, device=preds.device)
        self.scores.append(score.unsqueeze(0))  # keep batch dim

    def compute(self) -> torch.Tensor:
        # Aggregate across all batches — mean is typical, adjust as needed
        return dim_zero_cat(self.scores).mean()

    def reset(self) -> None:
        super().reset()  # ← critical: clears add_state'd variables
        # No need to manually clear self.scores — super().reset() does it
```

**Usage with MONAI Dice:**
```python
from monai.metrics import DiceMetric

monai_dice = DiceMetric(include_background=False, reduction="mean")

def monai_dice_fn(preds, target):
    # MONAI expects (N, C, H, W) one-hot format
    return monai_dice(preds, target).mean()

metric = WrappedExternalMetric(monai_dice_fn)
```

**Why the epoch-level mismatch happens** (from @ibro45's original plots): if you don't use `add_state`, TorchMetrics has no state to aggregate, so `.compute()` only reflects the last batch. The pattern above stores every batch's score and averages at compute time.

Docs: [Implementing Custom Metrics](https://lightning.ai/docs/torchmetrics/stable/pages/implement.html)
